plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/MemoryParquetDataSource.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.parquet;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ListMultimap;
+import io.airlift.slice.Slice;
+import io.trino.filesystem.TrinoInput;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.memory.context.AggregatedMemoryContext;
+import io.trino.memory.context.LocalMemoryContext;
+import io.trino.parquet.ChunkReader;
+import io.trino.parquet.DiskRange;
+import io.trino.parquet.ParquetDataSource;
+import io.trino.parquet.ParquetDataSourceId;
+import io.trino.parquet.reader.ChunkedInputStream;
+import io.trino.plugin.hive.FileFormatDataSourceStats;
+import jakarta.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class MemoryParquetDataSource
+        implements ParquetDataSource
+{
+    private final ParquetDataSourceId id;
+    private final long readTimeNanos;
+    private final long readBytes;
+    private final LocalMemoryContext memoryUsage;
+    @Nullable
+    private Slice data;
+
+    public MemoryParquetDataSource(TrinoInputFile inputFile, AggregatedMemoryContext memoryContext, FileFormatDataSourceStats stats)
+            throws IOException
+    {
+        try (TrinoInput input = inputFile.newInput()) {
+            long readStart = System.nanoTime();
+            this.data = input.readTail(toIntExact(inputFile.length()));
```

Comment:
In situations where, eg: the file size changed between the split enumeration on the coordinator and the worker read, we might fail to actually capture the entire file here. Instead, we should probably prefer not calling `readTail` but instead read from the file beginning and verify that we encounter an `EOF` on the stream so that we can handle that case explicitly (by failing, or falling back to the normal parquet datasource)

Reply:
> file size changed between the split enumeration on the coordinator and the worker read

Does that case work in the current code ? I see `readTail` calls in `MetadataReader#readFooter`, wouldn't these already fail if the file size changed ?